File: backend/app/models/field_submission.py
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from app.models.field import FieldType
from app.models.goal import GoalSide
from app.models.position import Position


class Dimensions(BaseModel):
    length: float = Field(gt=0)
    width: float = Field(gt=0)
    unit: Literal["cm"]


class SubmittedTeam(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    color: str = Field(pattern=r"^#[0-9A-Fa-f]{6}$")
    defended_goal_id: str = Field(alias="defendedGoalId", min_length=1)


class SubmittedGoal(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    side: GoalSide
    coordinates: tuple[Position, Position, Position, Position]


class SubmittedPlayer(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    number: int = Field(gt=0, le=99)
    team_id: str = Field(alias="teamId", min_length=1)
    position: Position
    orientation: float = Field(ge=0, lt=360)
    velocity: Position
    speed_category: Literal["BASELINE", "FAST", "SUPER_FAST"] = Field(
        default="BASELINE",
        alias="speedCategory",
    )


class SubmittedBall(BaseModel):
    position: Position
    direction: float = Field(ge=0, lt=360)
    speed: float = Field(ge=0)


class CircularSubmittedOpenSpace(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    type: Literal["circular"]
    center: Position
    radius: float = Field(gt=0)


class RectangularSubmittedOpenSpace(BaseModel):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    type: Literal["rectangular"]
    bottom_left: Position = Field(alias="bottomLeft")
    top_right: Position = Field(alias="topRight")


SubmittedOpenSpace = Annotated[
    CircularSubmittedOpenSpace | RectangularSubmittedOpenSpace,
    Field(discriminator="type"),
]


class SubmittedFieldConfiguration(BaseModel):
    label: str = Field(min_length=1)
    field_type: FieldType = Field(alias="fieldType")
    dimensions: Dimensions
    goal_dimensions: Dimensions = Field(alias="goalDimensions")
    teams: tuple[SubmittedTeam, SubmittedTeam]
    goals: tuple[SubmittedGoal, SubmittedGoal]
    players: list[SubmittedPlayer] = Field(default_factory=list)
    ball: SubmittedBall
    open_spaces: list[SubmittedOpenSpace] = Field(
        alias="openSpaces",
        default_factory=list,
    )

class FieldSubmission(BaseModel):
    schema_version: Literal["1.0"] = Field(alias="schemaVersion")
    tactical_instruction: str | None = Field(
        default=None,
        alias="tacticalInstruction",
        max_length=500,
    )
    field_configuration: SubmittedFieldConfiguration = Field(
        alias="fieldConfiguration"
    )

File: frontend/src/models/field-submission.ts
import { FieldConfiguration, FieldFormat } from "./field-configuration";
import {
  FIELD_LENGTH_CM,
  FIELD_WIDTH_CM,
  GOAL_LENGTH_CM,
  GOAL_WIDTH_CM,
} from "./field-coordinate";
import { Goal } from "./goal";
import { OpenSpace } from "./open-space";
import { Position } from "./position";
import { PlayerSpeedCategory } from "./player";
import { Team } from "./team";

export const FIELD_SUBMISSION_SCHEMA_VERSION = "1.0" as const;

export type SubmittedPlayer = {
  id: string;
  name: string;
  number: number;
  teamId: string;
  position: Position;
  orientation: number;
  velocity: Position;
  speedCategory: PlayerSpeedCategory;
};

export type FieldSubmission = {
  schemaVersion: typeof FIELD_SUBMISSION_SCHEMA_VERSION;
  tacticalInstruction?: string;
  fieldConfiguration: {
    label: string;
    fieldType: FieldFormat;
    dimensions: {
      length: number;
      width: number;
      unit: "cm";
    };
    goalDimensions: {
      length: number;
      width: number;
      unit: "cm";
    };
    teams: [Team, Team];
    goals: [Goal, Goal];
    players: SubmittedPlayer[];
    ball: {
      position: Position;
      direction: number;
      speed: number;
    };
    openSpaces: OpenSpace[];
  };
};

/** Builds the immutable JSON-compatible snapshot submitted to the backend. */
export function createFieldSubmission(
  configuration: FieldConfiguration,
  tacticalInstruction?: string,
): FieldSubmission {
  const normalizedInstruction = tacticalInstruction?.trim();
  return {
    schemaVersion: FIELD_SUBMISSION_SCHEMA_VERSION,
    ...(normalizedInstruction
      ? { tacticalInstruction: normalizedInstruction }
      : {}),
    fieldConfiguration: {
      label: configuration.label,
      fieldType: configuration.fieldType,
      dimensions: {
        length: FIELD_LENGTH_CM,
        width: FIELD_WIDTH_CM,
        unit: "cm",
      },
      goalDimensions: {
        length: GOAL_LENGTH_CM,
        width: GOAL_WIDTH_CM,
        unit: "cm",
      },
      teams: configuration.teams.map((team) => ({ ...team })) as [Team, Team],
      goals: configuration.goals.map((goal) => ({
        ...goal,
        coordinates: goal.coordinates.map((coordinate) => ({
          ...coordinate,
        })) as Goal["coordinates"],
      })) as [Goal, Goal],
      players: configuration.players.map((player) => ({
        id: player.id,
        name: player.name,
        number: player.number,
        teamId: player.teamId,
        position: { ...player.position },
        orientation: player.orientation,
        speedCategory: player.speedCategory,
        // The editor submits a static snapshot, not an in-progress animation.
        velocity: { x: 0, y: 0 },
      })),
      ball: {
        position: { ...configuration.ball.position },
        direction: configuration.ball.direction,
        speed: 0,
      },
      openSpaces: configuration.openSpaces.map((openSpace) =>
        openSpace.type === "circular"
          ? { ...openSpace, center: { ...openSpace.center } }
          : {
              ...openSpace,
              bottomLeft: { ...openSpace.bottomLeft },
              topRight: { ...openSpace.topRight },
            },
      ),
    },
  };
}
